File: python/src/aoc_2024_02.py
"""Advent of Code 2024 - Day 2."""

from itertools import pairwise
from sys import stdin


def is_safe(xs: list[int]) -> bool:
    """Test is the levels are safe.

    Safe if:
     * all increasing or all decreasing and all ga
     * all differences are between 1 and 3
    """
    increasing: bool = xs[1] > xs[0]
    for x, y in pairwise(xs):
        diff = abs(y - x)
        if (y > x) != increasing or diff < 1 or diff > 3:
            return False
    return True


def is_safe_one(xs: list[int]) -> bool:
    """Test is the levels are safe allowing for one level to be removed."""
    if is_safe(xs):
        return True
    for i in range(len(xs)):
        ys = xs[:i] + xs[i + 1 :]
        if is_safe(ys):
            return True
    return False


if __name__ == "__main__":
    lines: list[list[int]] = [[int(x) for x in s.split()] for s in stdin.readlines()]
    print(sum(is_safe(line) for line in lines))
    print(sum(is_safe_one(line) for line in lines))

File: rust/src/bin/aoc_2024_02.rs
// Advent of Code, 2024 day 02

use std::io;

fn parse(line: &str) -> Vec<i32> {
    line.split_ascii_whitespace()
        .map(|s| s.parse().unwrap())
        .collect()
}

fn is_safe(xs: &[i32]) -> bool {
    let first_increasing: bool = xs[1] > xs[0];
    for pair in xs.windows(2) {
        let diff = (pair[1] - pair[0]).abs();
        let increasing = pair[1] > pair[0];
        if increasing != first_increasing || !(1..=3).contains(&diff) {
            return false;
        }
    }
    true
}

fn is_safe_one(xs: &[i32]) -> bool {
    if is_safe(xs) {
        return true;
    }
    for i in 0..xs.len() {
        let mut ys: Vec<i32> = Vec::new();
        ys.extend_from_slice(&xs[..i]);
        ys.extend_from_slice(&xs[(i + 1)..]);
        if is_safe(&ys) {
            return true;
        }
    }
    false
}

fn main() {
    let lines: Vec<Vec<i32>> = io::stdin()
        .lines()
        .map(|line| parse(&line.unwrap()))
        .collect();

    println!("{}", lines.iter().filter(|line| is_safe(line)).count());
    println!("{}", lines.iter().filter(|line| is_safe_one(line)).count());
}
